components/common-go/process/process.go

```diff
@@ -14,3 +21,36 @@ func IsNotChildProcess(err error) bool {
 
 	return (err.Error() == "wait: no child processes" || err.Error() == "waitid: no child processes")
 }
+
+var ErrForceKilled = errors.New("process didn't temrinate, so we send SIGKILL")
+
+// TerminateSync sends a SIGTERM to the given process and returns when the process has terminated but at most the given gracePeriod
+// if the gracePeriod ends we send SIGKILL to the process and return immediately with ErrForceKilled
+func TerminateSync[T any](pid int, gracePeriodOver <-chan T) error {
+	process, err := os.FindProcess(pid)
+	if err != nil { // never happens on UNIX
+		return err
+	}
+	err = process.Signal(unix.SIGTERM)
+	if err == os.ErrProcessDone {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	terminated := make(chan error, 1)
+	go func() {
+		_, err := process.Wait()
+		terminated <- err
```

Comment:
If the process never stops, and we reach `gracePeriodOver` we will leak this go-routine and also the `terminated` channel.

What's the expected behaviour here? Is it that `process.Kill()` causes `process.Wait()` to unblock?

Reply:
I haven't thought of the leakage of the routine, but, yes, process.Kill() will unblock Wait().